### DISCUSSION TITLE
Inconsistent order for custom variants with abritrary groups

### DISCUSSION THREAD

**Discussion by royvanv** [tailwindlabs/tailwindcss#14924] [Help] [answered]
**What version of Tailwind CSS are you using?**
v3.4.13

**What build tool (or framework if it abstracts the build tool) are you using?**
Tailwind CLI with `--postcss` flag (postcss v8.4.47)

**What version of Node.js are you using?**
v18.18.2

**What browser are you using?**
N/A

**What operating system are you using?**
N/A

**Reproduction URL**

https://play.tailwindcss.com/Eu7Tz5Jcrc

Toggle the `target` state using your browser's dev tool, because using anchor links breaks the playground.

**Describe your issue**

When defining custom variants in a plugin and using them in combination with abritrary groups the order of the selector is different than that of default variants (such as `hover`).

Basic plugin that adds variant `target` with the pseudo selector `:target`

```
plugin(function ({ addVariant }) {
    addVariant('target', '&:target');
}),
```

Expected class structure
```
group-[.test-group]:hover:<utility-class>
group-[.test-group]:target:<utility-class>
```

Actual class structure
```
group-[.test-group]:hover:<utility-class>
target:group-[.test-group]:<utility-class>
```

**Comment by royvanv** [comment]
FYI: it also happens in v4.0.0-alpha.30, see https://play.tailwindcss.com/Eu7Tz5Jcrc

**Comment by adamwathan** [comment, marked as answer]
Hey sorry can you explain the problem more clearly? What is the visible behavior in the browser you are expecting to see and what are you seeing instead? Are you trying to style a child based on when the _group_ matches `:target`?

I'm not sure what you mean when you say "expected class structure" — can you show me the class you are writing and the expected CSS output?

When I test the behavior is as I'd expect, these two classes both trigger correctly for me:

```
group-[.test-group]:hover:<utility-class>
group-[.test-group]:target:<utility-class>
```

When I hover the child, the hover styles appear, and when I make the child match `:target`, those styles appear:

https://github.com/user-attachments/assets/3a75d20e-d502-40c8-a1aa-f738396d9bb4

I get the same behavior in both v3 and v4.

> the order of the selector is different than that of default variants

This is the part I don't understand especially — the order is whatever order you type, you are the one determining the order 🤔 Do you mean you only get the behavior you expect if you write it in one order vs. the other order? If so I just need to clearly understand the actual behavior you are seeing vs. expecting.

Are you wanting to style things when the _group_ itself is the `:target` or something?

**Reply by royvanv** [reply]
What I was doing is replace a class `no-js` on the HTML element with `js` using JavaScript. And then use the `group-[.test-group]:target:` classes as fallback to set utility classes when an anchor was clicked and no JavaScript is supported. So wanted to style the `:target` based on the condition that the group exists.

**Comment by royvanv** [comment]
I am sorry, this seems to not be a bug. I must have messed something up in my project. Both orders seems to work now.

BTW, any plans to include `target` out of the box? Or is it too niche?

**Reply by adamwathan** [reply]
Cool! We do include it already actually: https://tailwindcss.com/docs/hover-focus-and-other-states#target
